File: workbench/server/ops.py
"""运维页数据面(板块四·运维): 常驻进程/开机任务/刷新轮/存储四块状态 + 白名单动作。

红线适配: 全程只读取证(端口/netstat/schtasks/文件体积), 动作一律 subprocess
(启动=拉起 cli.py/omniroute, 停止=taskkill 端口占用 PID), 不直写三板块文件;
唯一落盘是 bat 侧自己的 data\\*_task.log。工作台自身只读展示(你在看页面=它在跑)。
"""

import json
import re
import shutil
import subprocess
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]          # 仓库根
DATA = ROOT / "data"

# 常驻进程清单: key → (显示名, 端口, 说明)
PROCS = {
    "serve":     ("数据站 sources serve", 8787, "板块一读侧供数( Bearer 鉴权 )"),
    "workbench": ("工作台 workbench", 8788, "板块四本页面所在进程"),
    "omniroute": ("OmniRoute 翻译网关", 20128, "翻译链免费位(muse/mimo)所在"),
}

# 开机/计划任务清单(刷新型=已有; 自启型=本次新增)
TASKS = ["aag-sources-refresh", "aag-xsurge-refresh",
         "aag-serve", "aag-workbench", "aag-omniroute"]

LOG_FILES = ["refresh_task.log", "xsurge_task.log", "serve_task.log",
             "workbench_task.log", "omniroute_task.log", r"serve\access.log"]


def _run(cmd: list, timeout: int = 15) -> tuple[int, str]:
    try:
        p = subprocess.run(cmd, capture_output=True, timeout=timeout)
        out = (p.stdout or b"").decode("mbcs", errors="ignore") + \
              (p.stderr or b"").decode("mbcs", errors="ignore")
        return p.returncode, out.strip()
    except Exception as e:
        return -1, f"{type(e).__name__}: {e}"


def _port_pid(port: int) -> int:
    """netstat -ano 找 LISTENING 占用 PID(0=未监听)。"""
    rc, out = _run(["netstat", "-ano"], timeout=10)
    for line in out.splitlines():
        if f":{port}" in line and "LISTENING" in line.upper():
            parts = line.split()
            if parts and parts[-1].isdigit():
                return int(parts[-1])
    return 0


def _proc_status() -> list:
    rows = []
    for key, (name, port, note) in PROCS.items():
        pid = _port_pid(port)
        rows.append({"key": key, "name": name, "port": port, "note": note,
                     "up": pid > 0, "pid": pid})
    return rows


def _task_status() -> list:
    rows = []
    for name in TASKS:
        rc, out = _run(["schtasks", "/query", "/tn", name, "/fo", "csv", "/nh"])
        if rc != 0 or '"' not in out:
            rows.append({"name": name, "registered": False})
            continue
        parts = [c.strip().strip('"') for c in out.splitlines()[0].split('","')]
        row = {"name": name, "registered": True,
               "next_run": parts[1] if len(parts) > 1 else "",
               "state": parts[2] if len(parts) > 2 else ""}
        rc2, xml = _run(["schtasks", "/query", "/tn", name, "/xml"])
        m = re.search(r"<Interval>PT(\d+)M</Interval>", xml or "")
        if m:
            row["interval_min"] = int(m.group(1))
        rows.append(row)
    return rows


def _last_round() -> dict:
    out = {"sources": None, "xsurge": None}
    try:
        d = json.loads((DATA / "serve" / "refresh.json").read_text(encoding="utf-8"))
        srcs = d.get("sources") or {}
        ok = sum(1 for s in srcs.values() if s.get("status") == "ok")
        fail = sum(1 for s in srcs.values() if s.get("status") not in ("ok", "empty"))
        out["sources"] = {"started": d.get("round_started_at"),
                          "finished": d.get("round_finished_at"),
                          "ok": ok, "fail": fail, "new": sum(int(s.get("new") or 0)
                                                             for s in srcs.values()
                                                             if isinstance(s, dict))}
    except Exception:
        pass
    try:                                                    # xsurge 日志末行是 JSON
        lines = (DATA / "xsurge_task.log").read_text(encoding="utf-8",
                                                     errors="ignore").strip().splitlines()
        for ln in reversed(lines[-20:]):
            ln = ln.strip()
            if ln.startswith("{"):
                j = json.loads(ln)
                out["xsurge"] = {k: j.get(k) for k in
                                 ("ok", "failed", "todo", "tracked",
                                  "translated", "translate_failed") if k in j}
                break
    except Exception:
        pass
    return out


def _storage() -> dict:
    files = []
    for rel in LOG_FILES:
        p = DATA / rel
        files.append({"name": rel.replace("\\", "/"),
                      "mb": round(p.stat().st_size / 1048576, 2) if p.exists() else None})
    db = DATA / "serve" / "items.db"
    du = shutil.disk_usage(DATA.anchor)
    return {"items_db_mb": round(db.stat().st_size / 1048576, 1) if db.exists() else None,
            "logs": files, "disk_free_gb": round(du.free / 1073741824, 1)}


def status_payload() -> dict:
    return {"procs": _proc_status(), "tasks": _task_status(),
            "last_round": _last_round(), "storage": _storage(),
            "at": time.strftime("%Y-%m-%d %H:%M:%S")}


# ── 动作(白名单) ─────────────────────────────────────────────────────────────

def _spawn(cli_args: list, log: str) -> tuple[bool, str]:
    """后台拉起 cli.py 子命令(脱离本进程, 关窗口, 日志落 data\\<log>)。"""
    DATA.mkdir(exist_ok=True)
    f = open(DATA / log, "a", encoding="utf-8")
    f.write(f"\n===== {time.strftime('%F %T')} ops 拉起: {' '.join(cli_args)} =====\n")
    f.flush()
    subprocess.Popen(["py", "-3.11", str(ROOT / "cli.py")] + cli_args,
                     cwd=str(ROOT), stdout=f, stderr=subprocess.STDOUT,
                     creationflags=getattr(subprocess, "CREATE_NO_WINDOW", 0))
    return True, f"已后台启动(cli.py {' '.join(cli_args)})"


def action(target: str, op: str) -> dict:
    try:
        return _action_inner(target, op)
    except Exception as e:
        return {"ok": False, "output": f"{type(e).__name__}: {e}"}


def _action_inner(target: str, op: str) -> dict:
    if target == "serve":
        if op == "start":
            if _port_pid(8787):
                return {"ok": True, "output": "8787 已在监听, 无需启动"}
            return dict(zip(("ok", "output"), _spawn(["sources", "serve"], "serve_task.log")))
        if op == "stop":
            pid = _port_pid(8787)
            if not pid:
                return {"ok": True, "output": "8787 本就没在跑"}
            rc, out = _run(["taskkill", "/F", "/PID", str(pid)])
            return {"ok": rc == 0, "output": out or f"taskkill pid={pid} rc={rc}"}
    elif target == "omniroute":
        # npm 装的 omniroute 是 .cmd shim, CreateProcess 不能直接执行, 必须经 cmd /c
        if op == "start":
            if _port_pid(20128):
                return {"ok": True, "output": "20128 已在监听, 无需启动"}
            import os
            env = dict(os.environ, OMNIROUTE_SERVER_HOST="127.0.0.1")
            p = subprocess.run(["cmd", "/c", "omniroute", "serve", "--daemon", "--no-open"],
                               capture_output=True, timeout=60, env=env)
            out = ((p.stdout or b"") + (p.stderr or b"")).decode("utf-8", errors="ignore")
            return {"ok": p.returncode == 0, "output": out[-400:]}
        if op == "stop":
            p = subprocess.run(["cmd", "/c", "omniroute", "stop"],
                               capture_output=True, timeout=30)
            out = ((p.stdout or b"") + (p.stderr or b"")).decode("utf-8", errors="ignore")
            return {"ok": p.returncode == 0, "output": out[-300:]}
    elif target == "refresh":
        if op == "start":                                # 立即刷一轮(幂等, 与任务计划并存)
            return dict(zip(("ok", "output"), _spawn(["sources", "refresh"], "refresh_task.log")))
    elif target == "xsurge":
        if op == "start":
            return dict(zip(("ok", "output"),
                            _spawn(["workbench", "refresh-x-surge"], "xsurge_task.log")))
    return {"ok": False, "output": f"不支持的动作: {target}/{op}"}

File: workbench/web/js/pages/ops.js
/* 运维页: 数据站 24/7 控制台——常驻进程 / 开机自启任务 / 最近刷新轮 / 存储与日志。
   状态 15s 轮询; 动作走 /wb-api/ops/action 白名单(后端 subprocess, 不直写三板块)。 */
window.WB = window.WB || {};
WB.pages = WB.pages || {};

WB.pages.ops = {
  data() {
    return { d: null, busy: "", err: "" };
  },
  methods: {
    async load() {
      try { this.d = await WB.api.get("/ops/status"); this.err = ""; }
      catch (e) { this.err = e.error || "状态读取失败"; }
    },
    async act(target, op, label) {
      if (this.busy) return;
      this.busy = target + op;
      try {
        const r = await WB.api.post("/ops/action", { target, op });
        WB.toast((r.ok ? "✓ " : "✗ ") + (r.output || label));
      } catch (e) { WB.toast("✗ " + (e.error || "动作失败")); }
      this.busy = "";
      setTimeout(this.load, 2500);          // 进程起停有秒级延迟, 2.5s 后刷新状态
    },
    fmtTask(t) {
      if (!t.registered) return "未注册";
      return t.state + (t.next_run ? " · 下次 " + t.next_run : "") +
             (t.interval_min ? " · 每" + t.interval_min + "分钟" : " · 登录时");
    },
  },
  mounted() { this.load(); this._t = setInterval(this.load, 15000); },
  unmounted() { clearInterval(this._t); },
  template: `
  <div style="max-width:820px">
    <div v-if="err" class="card"><span class="muted">✗ {{ err }}</span></div>
    <template v-if="d">

    <!-- 1. 常驻进程 -->
    <div class="card">
      <h3>常驻进程 <button class="btn" style="float:right" @click="load">刷新</button></h3>
      <table class="tbl">
        <thead><tr><th>进程</th><th>端口</th><th>状态</th><th>PID</th><th>操作</th></tr></thead>
        <tbody>
          <tr v-for="p in d.procs" :key="p.key">
            <td>{{ p.name }}<div class="muted" style="font-size:12px">{{ p.note }}</div></td>
            <td class="mono">{{ p.port }}</td>
            <td><span class="dot" :class="p.up ? 'ok' : 'err'"></span>
                {{ p.up ? '运行中' : '未运行' }}</td>
            <td class="mono">{{ p.pid || '—' }}</td>
            <td>
              <template v-if="p.key === 'workbench'">
                <span class="muted" style="font-size:12px">本页面所在进程</span>
              </template>
              <template v-else>
                <button class="btn" :disabled="busy" v-if="!p.up"
                        @click="act(p.key, 'start', '启动')">启动</button>
                <button class="btn" :disabled="busy" v-else
                        @click="act(p.key, 'stop', '停止')">停止</button>
                <button class="btn" :disabled="busy" v-if="p.up"
                        @click="act(p.key, 'stop', '重启'); act(p.key, 'start', '重启')">重启</button>
              </template>
            </td>
          </tr>
        </tbody>
      </table>
      <p class="muted" style="margin-top:8px">OmniRoute 未运行时翻译链自动落付费位(无缝), 无事故;
        桌面「数据站控制台」快捷方式可在全灭时一键拉起本页。</p>
    </div>

    <!-- 2. 开机自启与计划任务 -->
    <div class="card">
      <h3>任务计划</h3>
      <table class="tbl">
        <thead><tr><th>任务</th><th>状态</th></tr></thead>
        <tbody>
          <tr v-for="t in d.tasks" :key="t.name">
            <td class="mono">{{ t.name }}</td>
            <td :class="t.registered ? '' : 'muted'">
              <span class="dot" :class="t.registered ? 'ok' : 'err'"></span>
              {{ fmtTask(t) }}</td>
          </tr>
        </tbody>
      </table>
      <div class="form-row" style="margin-top:10px">
        <button class="btn" :disabled="busy" @click="act('refresh', 'start', '刷新一轮')">立即刷新一轮</button>
        <button class="btn" :disabled="busy" @click="act('xsurge', 'start', 'X采集一轮')">立即X采集一轮</button>
        <span class="muted">与计划任务并存, 幂等可重跑</span>
      </div>
    </div>

    <!-- 3. 最近刷新轮 -->
    <div class="card">
      <h3>最近刷新轮</h3>
      <template v-if="d.last_round.sources">
        <div class="form-row"><label>主站刷新</label>
          <span>{{ d.last_round.sources.started }} → {{ d.last_round.sources.finished }}</span></div>
        <div class="form-row"><label>结果</label>
          <span>{{ d.last_round.sources.ok }} 源成功 / {{ d.last_round.sources.fail }} 失败 /
                新入库 {{ d.last_round.sources.new }} 条</span></div>
      </template>
      <template v-if="d.last_round.xsurge">
        <div class="form-row"><label>X 起爆帖</label>
          <span class="mono" style="font-size:12px">{{ JSON.stringify(d.last_round.xsurge) }}</span></div>
      </template>
      <p v-if="!d.last_round.sources && !d.last_round.xsurge" class="muted">暂无轮次记录</p>
    </div>

    <!-- 4. 存储与日志 -->
    <div class="card">
      <h3>存储与日志</h3>
      <div class="form-row"><label>items.db</label>
        <span>{{ d.storage.items_db_mb }} MB</span>
        <span class="muted">保留窗自清: flash 48h / 其他 7d</span></div>
      <div class="form-row"><label>磁盘剩余</label><span>{{ d.storage.disk_free_gb }} GB</span></div>
      <table class="tbl" style="margin-top:6px">
        <thead><tr><th>日志</th><th>体积</th></tr></thead>
        <tbody>
          <tr v-for="f in d.storage.logs" :key="f.name">
            <td class="mono" style="font-size:12px">data/{{ f.name }}</td>
            <td>{{ f.mb == null ? '—' : f.mb + ' MB' }}</td>
          </tr>
        </tbody>
      </table>
    </div>

    <p class="muted" style="font-size:12px">状态每 15s 自动刷新 · {{ d.at }}</p>
    </template>
    <div v-else class="card"><span class="muted">读取中…</span></div>
  </div>`,
};
